### ISSUE TITLE
[BUG] Temp allocations using `get_default_resource()` flawed for multi-GPU per process applications 

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#5839]
**Describe the bug**

libcudf's policy for any API is to use the user-provided memory resource for allocating any device memory returned to the caller. Temporary allocations should always go through the resource returned from `get_default_resource()`. This is potentially flawed for users that are using libcudf on multiple GPUs in the same process. 

For example, imagine 2 GPUs being used in the same process by 2 different threads
```
// thread 0
cudaSetDevice(0);
auto pool0 = rmm::mr::pool_memory_resource{}; // Pool unique to dev0
auto result = cudf::some_function(..., &pool0); // returned memory allocated with pool0
```

```
// thread 1
cudaSetDevice(1);
auto pool1 = rmm::mr::pool_memory_resource{}; // Pool unique to dev1
auto result = cudf::other_function(..., &pool1); // returned memory allocated with pool1
```

Both threads create their own independent pools on their respective devices. The above example is correct because temporary allocations will use the `cuda_memory_resource` returned from `get_default_resource()`, however, it will be slow. Really, we want to use a pool for temporary allocations too. However, this isn't possible if the resource isn't device aware (e.g., like `cnmem_memory_resource`). Consider:

```
// thread 0
cudaSetDevice(0);
auto pool0 = rmm::mr::pool_memory_resource{}; // Pool unique to dev0
rmm::mr::set_default_resource(&pool); // Sets the default to use pool0
```

```
// thread 1
cudaSetDevice(1);
auto pool1 = rmm::mr::pool_memory_resource{}; // Pool unique to dev1
rmm::mr::set_default_resource(&pool1);  // sets the default to use pool1 
```

Now the two threads are conflicting in setting the default resource. There is only _one_ default resource, so only one of the thread's pools will be set. If one thread attempts to use the other thread's pool, this will fail spectacularly. Technically this could work with `cnmem_memory_resource` if constructed correctly because CNMeM is device aware, but we don't want to require all RMM resources to be device aware. 

Alternatively, you could use some form of mutual exclusion to ensure only one thread is active at a time and sets the default resource to it's pool, but this defeats the point of using multiple threads/devices concurrently in the first place. 

**Solution**

Use a device specific resource for temporary allocations, like `get_current_device_resource` from https://github.com/rapidsai/rmm/pull/458. This basically means replacing all current `get_default_resource` calls with `get_current_device_resource` in both RMM and libcudf.

**Comment by jrhemstad** [comment]
This was closed by #459.

**Comment by harrism** [comment]
I guess you mean rapidsai/rmm#459?
